MDEV-27342: Fix issue of recovery failure using new server id

HugoWenTD opened the pull request:
## Description
Commit 6c39eaeb1 made the crash recovery dependent on server_id.
The crash recovery could fail when restoring a new instance from
original crashed data directory USING A NEW SERVER ID.

The issue doesn't exist in previous major versions before 10.6.

Root cause is when generating the input XID to be searched in the hash,
server id is populated with the current server id.
So if the server id changed when recovering, the XID couldn't be found
in the hash due to server id doesn't match.

This fix is to use original server id when creating the input XID
object in function `xarecover_do_commit_or_rollback`.


## How can this PR be tested?
I've verified manually the issue was fixed with the commit.
Also a new MTR test binlog.binlog_xa_recover_using_new_server_id is added to verify the scenario.

## Basing the PR against the correct MariaDB version

- [x] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

## Copyright

All new code of the whole pull request, including one or several files that are either new files or modified ones, are contributed under the BSD-new license. I am contributing on behalf of my employer Amazon Web Services, Inc.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=1983) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=1983) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=1983) it.</sub>

dlenski commented:
> Root cause is when generating the input XID to be searched in the hash,
> server id is populated with the current server id.
> So if the server id changed when recovering, the XID couldn't be found
> in the hash due to server id doesn't match.

@HugoWenTD explains this in detail in [a comment on MDEV-27342](https://jira.mariadb.org/browse/MDEV-27342?focusedCommentId=210590#comment-210590), but it may mentioning again here for reviewers. The code that is comparing the XIDs (and failing without this change) is here:

https://github.com/MariaDB/server/blob/428b057ee06dca05a3d20f5b4cc32ceedbc37836/storage/innobase/trx/trx0trx.cc#L2003-L2014

andrelkin commented on sql/handler.cc:
```diff
@@ -2658,9 +2660,12 @@ static my_bool xarecover_handlerton(THD *unused, plugin_ref plugin,
           if (xid_member_replace(info->commit_list, x, info->mem_root,
-                                 is_server_xid? NULL : &info->list[i]))
+                                 is_server_xid? NULL : &info->list[i],
+                                 is_server_xid? info->list[i].get_trx_server_id(): server_id))
```
Right, we have to pass the xid's `server_id` down to  `xid_recovery_member` ctor, but I suggest to
pass zero (which would be the default of  `xid_recovery_member::server_id`) when `get_trx_server_id() == server_id`.
As the latter is the normal (It feels not so many users recover along your way :-))  case,  `xid_recovery_member::server_id == 0` would avoid `memcpy` in `xid_t::set`.

HugoWenTD replied:
Thank you! Good point! Updated in new commit.

andrelkin commented on sql/handler.cc:
```diff
@@ -2658,9 +2660,12 @@ static my_bool xarecover_handlerton(THD *unused, plugin_ref plugin,
         {
-          // remember "full" xid too when it's not in mysql format
+          // remember "full" xid too when it's not in mysql format.
+          // MDEV-27342 - Also record the server_id of the transaction. It will be used for
```
Please don't refer to MDEV:s. It just litters. If needed they can be always found with git annotate.

HugoWenTD replied:
Updated in commit.

andrelkin commented on sql/handler.h:
```diff
@@ -891,12 +891,13 @@ struct xid_t {
     formatID= 1;
     set(MYSQL_XID_PREFIX_LEN, 0, MYSQL_XID_PREFIX);
-    memcpy(data+MYSQL_XID_PREFIX_LEN, &server_id, sizeof(server_id));
+    memcpy(data+MYSQL_XID_PREFIX_LEN, &trx_server_id, sizeof(trx_server_id));
```
So this should be made optional to run only when `server_id != 0`.

andrelkin replied:
Sorry, this was my blunder: `memcpy` is always to be executed of course.

andrelkin commented on sql/handler.h:
```diff
@@ -963,11 +971,12 @@ struct xid_recovery_member
 
   xid_recovery_member(my_xid xid_arg, uint prepare_arg, bool decided_arg,
-                      XID *full_xid_arg)
+                      XID *full_xid_arg, ulong server_id_arg)
```
And the same `decltype(::server_id)` to replace `ulong` that you won't to explain why anymore.

andrelkin commented on sql/handler.h:
```diff
@@ -891,12 +891,13 @@ struct xid_t {
   }
-  void set(ulonglong xid)
+  // MDEV-27342 - Populate server_id if it's specified, otherwise use the current server_id
+  void set(ulonglong xid, ulong trx_server_id= server_id)
```
s/ulong/decltype/ - please read on.

andrelkin commented on sql/handler.h:
```diff
@@ -963,11 +971,12 @@ struct xid_recovery_member
   bool decided_to_commit;
   Binlog_offset binlog_coord; // semisync recovery binlog offset
   XID *full_xid;           // needed by wsrep or past it recovery
+  ulong server_id;         // server id of orginal server
```
While I was assuming (before I checked indeed) the new member is of the same type as the global var of `server_id`,
the declaration like
`decltype(::server_id) server_id;`
would speak it the most explicit way.

HugoWenTD replied:
Replaced all ulong to decltype in commit.

andrelkin commented on sql/handler.h:
```diff
@@ -963,11 +971,12 @@ struct xid_recovery_member
+                      XID *full_xid_arg, ulong server_id_arg)
     : xid(xid_arg), in_engine_prepare(prepare_arg),
-      decided_to_commit(decided_arg), full_xid(full_xid_arg) {};
+      decided_to_commit(decided_arg), full_xid(full_xid_arg) , server_id(server_id_arg) {};
```
The patch looks generally good.
I offered some optimization along with few minor notes.

andrelkin commented on sql/handler.cc:
```diff
@@ -2658,9 +2660,13 @@ static my_bool xarecover_handlerton(THD *unused, plugin_ref plugin,
         {
-          // remember "full" xid too when it's not in mysql format
+          // remember "full" xid too when it's not in mysql format.
+          // Also record the server_id of the transaction. It will be used for
```
I think a substitution like
  s/the server_id of the transaction /the transaction's original server_id if it's different from the current server_id/
reads better. Feel free to polish it, just two server_id:s should be put in contrast here.

andrelkin reviewed (commented):
Changes, including a test, are requested.

HugoWenTD commented:
> Changes, including a test, are requested.

In addition, seems two tests failed after making default value to 0:
innodb.group_commit_crash 'innodb'       w2 [ fail ]
innodb.group_commit_crash_no_optimize_thread 'innodb' w3 [ fail ]

I'll look into it later and add test as well. Thanks!

HugoWenTD commented:
@andrelkin Thank you for the suggestions! I've updated the commit. Please see my reply to some of the comments below:

---
Added a new test `binlog.binlog_xa_recover_using_new_server_id` as requested.

---

After further consideration, I found that making the default value to 0 when `xid_member_replace` made the conditions a little complex.
For example in `set` function need to check the input orig server_id first and memcpy if it's not 0, otherwise also need to check current server_id and memcpy if it's not 0.
(just like a few test failed because of not handling the conditions well in my last commit..)

So I think to make it simpler, maybe in this commit will just
- Add a parameter in `set` with default value using current server id.
```
   void set(ulonglong xid, decltype(::server_id) trx_server_id= server_id)
```
- And use original server_id as the input only when populating `xid_recovery_member`.

In this way we only changed the single scenario needed for `xid_recovery_member` but keep all existing behavior without the risk of breaking anything.

---

You mentioned "`xid_recovery_member::server_id == 0` would avoid memcpy in `xid_t::set`.".
I did the change in last commit but then I realized it's risky, and could cause issues because the server_id section is not initialized.
It's because `char data[XIDDATASIZE];` is not initialized in the [xid_t constructor](https://github.com/MariaDB/server/blob/10.6/sql/handler.h#L879) (no memset(0)), and it depends on the [set functions](https://github.com/MariaDB/server/blob/10.6/sql/handler.h#L884-L910) to populate it. Otherwise the server_id could be any uninitialized random value.
```
memcpy(data+MYSQL_XID_PREFIX_LEN, &server_id, sizeof(server_id));
```

One way is to use memset(0) if server_id is 0 in current `set` function. - but it may not make much difference than it is now.
Or even add the memset in constructor so all new object will init it - but that could cause performance regression as later all set functions will update it again.

Overall for now I suppose to keep it as is if you also agree.

---

Thanks for your review again!
Hugo

HugoWenTD commented on mysql-test/suite/binlog/t/binlog_xa_recover_using_new_server_id.test:
```diff
@@ -0,0 +1,51 @@
+
+--echo ========= Crash the server.
+--write_file $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+wait-binlog_xa_recover.test
```
Updated. Sorry I forgot to update the suffix, thanks for pointing out!

HugoWenTD commented on mysql-test/suite/binlog/t/binlog_xa_recover_using_new_server_id.test:
```diff
@@ -0,0 +1,51 @@
+
+--source include/have_innodb.inc
+--source include/have_debug.inc
+--source include/have_debug_sync.inc
```
updated

HugoWenTD commented on mysql-test/suite/binlog/t/binlog_xa_recover_using_new_server_id.test:
```diff
@@ -0,0 +1,51 @@
+--source include/not_valgrind.inc
+
+# (We do not need to restore these settings, as we crash the server).
+SET GLOBAL max_binlog_size= 4096;
```
updated and removed the settings .

andrelkin commented:
@ottok : the last commit is good. Thank you and especially :-) @HugoWenTD for persistence and creativity!

andrelkin commented:
@ottok: To how we proceed now, I am guessing that the pushing to main 10.6 is by myself who the owner of [MDEV-27342](https://jira.mariadb.org/browse/MDEV-27342). Confirmed?

HugoWenTD commented:
@andrelkin I just rebased the commit on latest 10.6 and now all builds passed. I think you could merge if no other concerns? Thank you very much!

andrelkin commented:
> @andrelkin I just rebased the commit on latest 10.6 and now all builds passed. I think you could merge if no other concerns? Thank you very much!

Thank you, Hugo! I am ushering the patch into the main trees.

andrelkin commented:
@ottok : I need instructions for that. There's no such button in my sight, nor text search helps to find it. (Could it be my firefox got crazy... too). Thank you and twice if you'll do that for me (I anyway have to learn doing that).

haidong commented:
I suspect the inability to merge was caused by `Some checks haven't completed yet`. Perhaps the `Contributor License Agreement` needs to be signed first?

HugoWenTD commented:
@haidong, I'm contributing under the `BSD-new license` so I think I shouldn't sign the `Contributor License Agreement` . 

However I'm not sure if that's one requirement for showing the `merge` button, as I didn't need to do that in previous pull requests for MariaDB. e.g. https://github.com/MariaDB/server/pull/1932

haidong commented:
> However I'm not sure if that's one requirement for showing the `merge` button, as I didn't need to do that in previous pull requests for MariaDB. e.g. #1932

Good point. Seeing #1932 was merged successfully by Daniel Black, could you look into this when you have a chance, @grooverdan ?

ottok commented:
@andrelkin There was this button, I pressed it now on your behalf.
![image](https://user-images.githubusercontent.com/668724/158297032-cf8d4f26-c497-4d86-8662-c9a5db5508ea.png)
